csrc/id_model/indexing.cpp

```diff
@@ -97,128 +98,214 @@ Val* getStrideOfGlobalMemoryTensor(TensorView* tv, int64_t alloc_dim) {
-  std::vector<Val*> actual_strides;
-  for (const auto i : c10::irange(allocation_domains.size())) {
-    auto allocation_domain = allocation_domains.at(i);
-    auto promotion_domain = promoted_allocation_domains.at(i);
-    if (!mayRequireAllocation(tv, allocation_domain)) {
-      continue;
+    // Compute the strides from innermost to outermost domains
+    std::vector<Val*> strides(allocation_domains.size(), nullptr);
+    Val* cur_contig_stride = tv->fusion()->oneVal();
+    for (const auto i : c10::irange(allocation_domains.size())) {
+      auto dim = allocation_domains.size() - i - 1;
+      auto allocation_domain = allocation_domains.at(dim);
+      auto promotion_domain = promoted_allocation_domains.at(dim);
+
+      if (!mayRequireAllocation(tv, allocation_domain)) {
+        continue;
+      }
+
+      const std::optional<bool> contig_flag = contiguity.at(dim);
+      // Broadcast doesn't have contig flag but it must have been
+      // already filtered out
+      NVF_ERROR(contig_flag.has_value());
+
+      if (contig_flag.value()) {
+        strides[dim] = cur_contig_stride;
+        cur_contig_stride = SimplifyingIrBuilder::mulExpr(
+            cur_contig_stride, promotion_domain->extent());
+      } else {
+        // Assume that the tensor should always be a Global memory
+        // tensor if it has non-contig allocation domains
+        NVF_ERROR(tv->getMemoryType() == MemoryType::Global);
+        strides[dim] = getStrideOfGlobalMemoryTensor(tv, (int64_t)dim);
+        cur_contig_stride = SimplifyingIrBuilder::mulExpr(
+            strides[dim], promotion_domain->extent());
+      }
+    }
+
+    // Filter out non-allocated domains. This is already done for Local
+    // and Shared tensors with no set allocation domains, but not for
+    // the other cases. For example, a reduction output tensor that is
+    // also a fusion output may still have reduction domains in their
+    // allocation domains, which aren't relevant for indexing
+    std::vector<IterDomain*> actual_allocation_domains;
+    std::vector<Val*> actual_strides;
+    for (const auto i : c10::irange(allocation_domains.size())) {
+      auto allocation_domain = allocation_domains.at(i);
+      auto promotion_domain = promoted_allocation_domains.at(i);
+      if (!mayRequireAllocation(tv, allocation_domain)) {
+        continue;
+      }
+      auto stride = strides.at(i);
+      NVF_ERROR(stride != nullptr);
+      actual_allocation_domains.push_back(promotion_domain);
+      actual_strides.push_back(stride);
     }
-    auto stride = strides.at(i);
-    NVF_ERROR(stride != nullptr);
-    actual_allocation_domains.push_back(promotion_domain);
-    actual_strides.push_back(stride);
+
+    return IndexingAllocationInfo{actual_allocation_domains, actual_strides};
   }
 
-  return {actual_allocation_domains, actual_strides};
-}
+  std::unordered_map<TensorView*, IndexingAllocationInfo> tv_alloc_info_map;
+  std::unordered_set<TensorView*> used_as_producer;
```

Comment:
Nit: our convention of appending underscore for private members.
```suggestion
  std::unordered_map<TensorView*, IndexingAllocationInfo> tv_alloc_info_map_;
  std::unordered_set<TensorView*> used_as_producer_;
```

Reply:
They are actually not private. This class was meant to be a relatively simple class, although it isn't anymore.